File: python/package/robot_simulator/robot_6DOF.py
import math
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

class robot_6DOF:

    d1 = 89.2 / 250
    d4 = 109.3 / 250
    d5 = 94.75 / 250
    d6 = 82.5 / 250
    a2 = -425 / 250
    a3 = -392 / 250
    alpha1 = np.pi / 2
    alpha4 = np.pi / 2
    alpha5 = -np.pi / 2

    def __init__(self, name, bias):
        self.name = name
        self.biasx = bias[0]
        self.biasy = bias[1]
        self.biasz = bias[2]

    def fkine(self, q):
        q = np.array(q).reshape(1, 6)
        q1 = q[0, 0]
        q2 = q[0, 1]
        q3 = q[0, 2]
        q4 = q[0, 3]
        q5 = q[0, 4]
        q6 = q[0, 5]

        A0 = np.array([[1, 0, 0, self.biasx],
                       [0, 1, 0, self.biasy],
                       [0, 0, 1, self.biasz],
                       [0, 0, 0, 1]])

        A1 = np.array([[math.cos(q1), 0,  math.sin(q1), 0],
                       [math.sin(q1), 0, -math.cos(q1), 0],
                       [0, 1, 0, self.d1],
                       [0, 0, 0, 1]])

        A2 = np.array([[math.cos(q2), -math.sin(q2), 0, self.a2 * math.cos(q2)],
                       [math.sin(q2),  math.cos(q2), 0, self.a2 * math.sin(q2)],
                       [0, 0, 1, 0],
                       [0, 0, 0, 1]])

        A3 = np.array([[math.cos(q3), -math.sin(q3), 0, self.a3 * math.cos(q3)],
                       [math.sin(q3),  math.cos(q3), 0, self.a3 * math.sin(q3)],
                       [0, 0, 1, 0],
                       [0, 0, 0, 1]])

        A4 = np.array([[math.cos(q4), 0,  math.sin(q4), 0],
                       [math.sin(q4), 0, -math.cos(q4), 0],
                       [0, 1, 0, self.d4],
                       [0, 0, 0, 1]])

        A5 = np.array([[math.cos(q5), 0, -math.sin(q5), 0],
                       [math.sin(q5), 0,  math.cos(q5), 0],
                       [0, -1, 0, self.d5],
                       [0,  0, 0, 1]])

        A6 = np.array([[math.cos(q6), -math.sin(q6), 0, 0],
                       [math.sin(q6),  math.cos(q6), 0, 0],
                       [0, 0, 1, self.d6],
                       [0, 0, 0, 1]])

        T0 = A0
        T1 = np.matmul(T0, A1)
        T2 = np.matmul(T1, A2)
        T3 = np.matmul(T2, A3)
        T4 = np.matmul(T3, A4)
        T5 = np.matmul(T4, A5)
        T6 = np.matmul(T5, A6)

        return T6, T5, T4, T3, T2, T1, T0


    def ikine(self, T):
        a = np.array([0, self.a2, self.a3, 0, 0, 0])
        d = np.array([self.d1, 0, 0, self.d4, self.d5, self.d6])
        alpha = np.array([np.pi/2, 0, 0, np.pi/2, -np.pi/2, 0])
        ##
        theta = np.zeros([8, 6])
        theta1 = np.zeros([1, 2])
        theta5 = np.zeros([2, 2])
        theta6 = np.zeros([2, 2])
        theta3 = np.zeros([4, 2])
        #

        nx = T[0][0]
        ny = T[1][0]
        nz = T[2][0]
        ox = T[0][1]
        oy = T[1][1]
        oz = T[2][1]
        ax = T[0][2]
        ay = T[1][2]
        az = T[2][2]
        px = T[0][3] - self.biasx
        py = T[1][3] - self.biasy
        pz = T[2][3] - self.biasz

        ## sovle joint1
        m = d[5] * ay - py
        n = ax * d[5] - px
        theta1[0][0] = math.atan2(m, n) - math.atan2(d[3], math.sqrt(math.pow(m, 2) + math.pow(n, 2) - math.pow(d[3], 2)))
        theta1[0][1] = math.atan2(m, n) - math.atan2(d[3], -math.sqrt(math.pow(m, 2) + math.pow(n, 2) - math.pow(d[3], 2)))

        ## solve joint5
        theta5[0, 0:2] = np.arccos(ax * np.sin(theta1) - ay * np.cos(theta1))
        theta5[1, 0:2] = -np.arccos(ax * np.sin(theta1) - ay * np.cos(theta1))

        ## solve joint6
        mm = nx * np.sin(theta1) - ny * np.cos(theta1)
        nn = ox * np.sin(theta1) - oy * np.cos(theta1)
        theta6 = np.arctan2(mm, nn) - np.arctan2(np.sin(theta5), 0)

        ## solve joint3
        mmm = d[4] * (np.sin(theta6) * (nx * np.cos(theta1) + ny * np.sin(theta1)) + np.cos(theta6) * (ox * np.cos(theta1) + oy * np.sin(theta1))) \
            - d[5] * (ax * np.cos(theta1) + ay * np.sin(theta1)) + px * np.cos(theta1) + py * np.sin(theta1)

        nnn = pz - d[0] - az * d[5] + d[4] * (oz * np.cos(theta6) + nz * np.sin(theta6))
        theta3[0:2, :] = np.arccos((np.power(mmm, 2) + np.power(nnn, 2) - np.square(a[1]) - np.square(a[2])) / (2 * a[1] * a[2]))
        theta3[2:4, :] = -np.arccos((np.power(mmm, 2) + np.power(nnn, 2) - np.square(a[1]) - np.square(a[2])) / (2 * a[1] * a[2]))

        ## solve joint2
        mmm_s2 = np.vstack((mmm, mmm))
        nnn_s2 = np.vstack((nnn, nnn))
        s2 = ((a[2] * np.cos(theta3) + a[1]) * nnn_s2 - a[2] * np.sin(theta3) * mmm_s2) / (np.power(a[1], 2) + np.power(a[2], 2) + 2 * a[1] * a[2] * np.cos(theta3))
        c2 = (mmm_s2 + a[2] * np.sin(theta3) * s2) / (a[2] * np.cos(theta3) + a[1])
        theta2 = np.arctan2(s2, c2)

        ## selltle joint angle
        theta[0:4, 0] = theta1[0, 0]
        theta[4:8, 0] = theta1[0, 1]
        theta[:, 1] = np.transpose([theta2[0, 0], theta2[2, 0], theta2[1, 0], theta2[3, 0], theta2[0, 1], theta2[2, 1], theta2[1, 1], theta2[3, 1]])
        theta[:, 2] = np.transpose([theta3[0, 0], theta3[2, 0], theta3[1, 0], theta3[3, 0], theta3[0, 1], theta3[2, 1], theta3[1, 1], theta3[3, 1]])
        theta[0: 2, 4] = theta5[0, 0]
        theta[2: 4, 4] = theta5[1, 0]
        theta[4: 6, 4] = theta5[0, 1]
        theta[6: 8, 4] = theta5[1, 1]
        theta[0: 2, 5] = theta6[0, 0]
        theta[2: 4, 5] = theta6[1, 0]
        theta[4: 6, 5] = theta6[0, 1]
        theta[6: 8, 5] = theta6[1, 1]

        ## solve joint4
        theta[:, 3] = np.arctan2(-np.sin(theta[:, 5]) * (nx * np.cos(theta[:, 0]) + ny * np.sin(theta[:, 0])) - np.cos(theta[:, 5]) * \
                                 (ox * np.cos(theta[:, 0]) + oy * np.sin(theta[:, 0])), oz * np.cos(theta[:, 5]) + nz * np.sin(theta[:, 5])) - theta[:, 1] - theta[:, 2]

        return np.array(theta)


    def plot(self, q):
        T6, T5, T4, T3, T2, T1, T0 = self.fkine(q)
        ##
        ax = plt.axes(projection='3d')
        # ax1 = Axes3D(fig)
        ax.plot([T0[0][3], T1[0][3], T2[0][3], T3[0][3], T4[0][3], T5[0][3], T6[0][3]],
                [T0[1][3], T1[1][3], T2[1][3], T3[1][3], T4[1][3], T5[1][3], T6[1][3]],
                [T0[2][3], T1[2][3], T2[2][3], T3[2][3], T4[2][3], T5[2][3], T6[2][3]],
                color='black', marker='*', linewidth=2)

        limit = 5 / 1
        ax.set_xlim3d(-limit, limit)
        ax.set_ylim3d(-limit, limit)
        ax.set_zlim3d(-limit, limit)

        return ax

    def circle(self, r):
        N = 10
        theta = np.linspace(0, np.pi, N)
        phi = np.linspace(0, 2*np.pi, N)

        x = np.zeros((N ** 2, 1))
        y = np.zeros((N ** 2, 1))
        z = np.zeros((N ** 2, 1))
        traj = np.zeros((N ** 2, 3))

        for i in range(N):
            for j in range(N):
                x[(N-1) * i + j] = r * np.sin(theta[i]) * np.cos(phi[j]) + self.biasx
                y[(N-1) * i + j] = r * np.sin(theta[i]) * np.sin(phi[j]) + self.biasy
                z[(N-1) * i + j] = r * np.cos(theta[i]) + self.biasz

                traj[(N-1) * i + j, 0:3] = np.array([x[(N-1) * i + j], y[(N-1) * i + j], z[(N-1) * i + j]]).reshape(1, 3)

        return traj

File: matlab/package/robot_simulator_3D/programme_modified/robot_6DOF.m
classdef robot_6DOF
    %UNTITLED Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
        biasx;
        biasy;
        biasz;
        d1 =  89.2/250;
        d4 = 109.3/250;
        d5 = 94.75/250;
        d6 =  82.5/250;
        a2 =  -425/250;
        a3 =  -392/250;
        alpha1 = pi/2;
        alpha4 = pi/2;
        alpha5 = -pi/2;
    end
    
    methods
        function obj = robot_6DOF(biasx,biasy,biasz)
            %UNTITLED Construct an instance of this class
            %   Detailed explanation goes here
            obj.biasx = biasx;
            obj.biasy = biasy;
            obj.biasz = biasz;
        end
    
        function [T6,T5,T4,T3,T2,T1,T0] = fkine(obj,q)
           %% initial parameter
            q1 = q(1);q2 = q(2);q3 = q(3);q4 = q(4);q5 = q(5);q6 = q(6);          
            bias.x = obj.biasx;
            bias.y = obj.biasy;
            bias.z = obj.biasz;
           %%
            A0 = [1 0 0 bias.x;
                  0 1 0 bias.y;
                  0 0 1 bias.z;
                  0 0 0 1];
            A1 = [cos(q1) 0  sin(q1) 0;
                  sin(q1) 0 -cos(q1) 0;
                  0       1        0 obj.d1;
                  0       0        0 1];
            A2 = [cos(q2) -sin(q2) 0 obj.a2*cos(q2);
                  sin(q2)  cos(q2) 0 obj.a2*sin(q2);
                  0        0       1 0;
                  0        0       0 1];
            A3 = [cos(q3) -sin(q3) 0 obj.a3*cos(q3);
                  sin(q3)  cos(q3) 0 obj.a3*sin(q3);
                  0        0       1 0;
                  0        0       0 1];
            A4 = [cos(q4) 0  sin(q4) 0;
                  sin(q4) 0 -cos(q4) 0;
                  0       1  0       obj.d4;
                  0       0  0       1];
            A5 = [cos(q5) 0 -sin(q5) 0;
                  sin(q5) 0  cos(q5) 0;
                  0      -1  0       obj.d5;
                  0       0  0       1];
            A6 = [cos(q6) -sin(q6) 0 0;
                  sin(q6)  cos(q6) 0 0;
                  0        0       1 obj.d6;
                  0        0       0 1];
            T6 = A0*A1*A2*A3*A4*A5*A6;
            T5 = A0*A1*A2*A3*A4*A5;
            T4 = A0*A1*A2*A3*A4;
            T3 = A0*A1*A2*A3;
            T2 = A0*A1*A2;
            T1 = A0*A1;
            T0 = A0; 
        end
            
        function q = ikine(obj,T)
            a = [0,obj.a2,obj.a3,0,0,0];
            d = [obj.d1,0,0,obj.d4,obj.d5,obj.d6];
            alpha = [pi/2,0,0,pi/2,-pi/2,0];
            nx = T(1,1);ny = T(2,1);nz = T(3,1);
            ox = T(1,2);oy = T(2,2);oz = T(3,2);
            ax = T(1,3);ay = T(2,3);az = T(3,3);
            px = T(1,4) - obj.biasx;
            py = T(2,4) - obj.biasy;
            pz = T(3,4) - obj.biasz;
    
            %求解关节角1
            m = d(6)*ay-py;  
            n = ax*d(6)-px; 
            theta1(1,1) = atan2(m,n)-atan2(d(4),sqrt(m^2+n^2-(d(4))^2));
            theta1(1,2) = atan2(m,n)-atan2(d(4),-sqrt(m^2+n^2-(d(4))^2));
  
            %求解关节角5
            theta5(1,1:2) = acos(ax*sin(theta1)-ay*cos(theta1));
            theta5(2,1:2) = -acos(ax*sin(theta1)-ay*cos(theta1));      

            %求解关节角6
            mm = nx*sin(theta1)-ny*cos(theta1); 
            nn = ox*sin(theta1)-oy*cos(theta1);
            theta6 = atan2(mm,nn)-atan2(sin(theta5),0);

            %求解关节角3
            mmm = d(5)*(sin(theta6).*(nx*cos(theta1)+ny*sin(theta1))+cos(theta6).*(ox*cos(theta1)+oy*sin(theta1))) ...
                  -d(6)*(ax*cos(theta1)+ay*sin(theta1))+px*cos(theta1)+py*sin(theta1);
            nnn = pz-d(1)-az*d(6)+d(5)*(oz*cos(theta6)+nz*sin(theta6));
            theta3(1:2,:) =  acos((mmm.^2+nnn.^2-(a(2))^2-(a(3))^2)/(2*a(2)*a(3)));
            theta3(3:4,:) = -acos((mmm.^2+nnn.^2-(a(2))^2-(a(3))^2)/(2*a(2)*a(3)));

            %求解关节角2
            mmm_s2(1:2,:) = mmm;
            mmm_s2(3:4,:) = mmm;
            nnn_s2(1:2,:) = nnn;
            nnn_s2(3:4,:) = nnn;
            s2 = ((a(3)*cos(theta3)+a(2)).*nnn_s2-a(3)*sin(theta3).*mmm_s2)./ ...
                ((a(2))^2+(a(3))^2+2*a(2)*a(3)*cos(theta3));
            c2 = (mmm_s2+a(3)*sin(theta3).*s2)./(a(3)*cos(theta3)+a(2));
            theta2 = atan2(s2,c2);   
    
            %整理关节角1 5 6 3 2
            theta(1:4,1) = theta1(1,1);theta(5:8,1)=theta1(1,2);
            theta(:,2) = [theta2(1,1),theta2(3,1),theta2(2,1),theta2(4,1),theta2(1,2),theta2(3,2),theta2(2,2),theta2(4,2)]';
            theta(:,3) = [theta3(1,1),theta3(3,1),theta3(2,1),theta3(4,1),theta3(1,2),theta3(3,2),theta3(2,2),theta3(4,2)]';
            theta(1:2,5) = theta5(1,1);theta(3:4,5)=theta5(2,1);
            theta(5:6,5) = theta5(1,2);theta(7:8,5)=theta5(2,2);
            theta(1:2,6) = theta6(1,1);theta(3:4,6)=theta6(2,1);
            theta(5:6,6) = theta6(1,2);theta(7:8,6)=theta6(2,2); 
    
            %求解关节角4
            theta(:,4) = atan2(-sin(theta(:,6)).*(nx*cos(theta(:,1))+ny*sin(theta(:,1)))-cos(theta(:,6)).* ...
                        (ox*cos(theta(:,1))+oy*sin(theta(:,1))),oz*cos(theta(:,6))+nz*sin(theta(:,6)))-theta(:,2)-theta(:,3);  
            q = theta;
        end
        
        function handle = plot(obj,q)
        %% initial parameter
            [T6,T5,T4,T3,T2,T1,T0] = fkine(obj,q);
           %% plot figure
            handle = plot3([T0(1,4) T1(1,4) T2(1,4) T3(1,4) T4(1,4) T5(1,4) T6(1,4)],...
                           [T0(2,4) T1(2,4) T2(2,4) T3(2,4) T4(2,4) T5(2,4) T6(2,4)],...
                           [T0(3,4) T1(3,4) T2(3,4) T3(3,4) T4(3,4) T5(3,4) T6(3,4)],'k-*','linewidth',2);
%             xlabel('x','fontsize',15)
%             ylabel('y','fontsize',15)
%             zlabel('z','fontsize',15)
           %% axis limitation
            axis([-5 5 -5 5 -5 5])
%             axis([0 5 0 5 0 5])
            daspect([1 1 1])
        end
        
        function [traj,handle] = circle(obj,r)
            bias.x = obj.biasx;
            bias.y = obj.biasy;
            bias.z = obj.biasz;
            N = 10;
            theta = linspace(0,pi,N);
            phi = linspace(0,2*pi,N);
            for i=1:N
                for j=1:N
                    x(N*(i-1) + j) = r * sin(theta(i))*cos(phi(j)) + bias.x;
                    y(N*(i-1) + j) = r * sin(theta(i))*sin(phi(j)) + bias.y;
                    z(N*(i-1) + j) = r * cos(theta(i)) + bias.z;
                    traj(N*(i-1) + j,:) = [x(N*(i-1) + j) y(N*(i-1) + j) z(N*(i-1) + j)];
                end
            end
            handle = plot3(x,y,z,'k*','linewidth',1);
        end
        
        function [qt,qdt,qddt] = jtraj(obj, q0, q1, tv, qd0, qd1)
            if length(tv) > 1
                tscal = max(tv);
                t = tv(:)/tscal;
            else
                tscal = 1;
                t = (0:(tv-1))'/(tv-1); % normalized time from 0 -> 1
            end

            q0 = q0(:);
            q1 = q1(:);

            if nargin == 4
                qd0 = zeros(size(q0));
                qd1 = qd0;
            elseif nargin == 6
                qd0 = qd0(:);
                qd1 = qd1(:);
            else
                error('incorrect number of arguments')
            end

            % compute the polynomial coefficients
            A = 6*(q1 - q0) - 3*(qd1+qd0)*tscal;
            B = -15*(q1 - q0) + (8*qd0 + 7*qd1)*tscal;
            C = 10*(q1 - q0) - (6*qd0 + 4*qd1)*tscal;
            E = qd0*tscal; % as the t vector has been normalized
            F = q0;

            tt = [t.^5 t.^4 t.^3 t.^2 t ones(size(t))];
            c = [A B C zeros(size(A)) E F]';

            qt = tt*c;

            % compute optional velocity
            if nargout >= 2
                c = [ zeros(size(A)) 5*A 4*B 3*C  zeros(size(A)) E ]';
                qdt = tt*c/tscal;
            end

            % compute optional acceleration
            if nargout == 3
                c = [ zeros(size(A))  zeros(size(A)) 20*A 12*B 6*C  zeros(size(A))]';
                qddt = tt*c/tscal^2;
            end
        end
        
    end
end

